### ISSUE TITLE
EASGD implementation guildance

### ISSUE THREAD

**Issue by Godricly** [dmlc/mxnet#1657]
Hi all,
 Now I'm trying to implement [EASGD](http://arxiv.org/pdf/1412.6651v8.pdf) using MXNet and give it a test. So I'm asking for some advice on its implementation. What I'm trying now is to modify _update_params_on_kvstore in model.py. I need to keep both the local model and a server averaging model. But it would be nasty if I'm going further for EAMSGD with momentum as momentum is on local. Is there any DIY instruction on optimizer to handl this? Thanks.

**Comment by piiswrong** [comment]
@mli

**Comment by Godricly** [comment]
Ok. So here are some updates and questions after I tried implementing EASGD and EAMSGD (momentum EASGD) for single machine case. Anyone interested in implementing this one can have an idea on how to do so. Right now the result looks good. I did't post the code here because I went the nasty approach of modifying model.py and optimizer.py files. 

auther implementation for Torch link: [mpiT](https://github.com/sixin-zh/mpiT)

How to implement in MxNet:

1.  modify the _update_params_on_kvstore and implement everything according to paper,I used the kvstore to manage averaging weight, and you need to pass in optimizer to learning rate etc..

2. create a dummy optimizer which passes pushed gradient to weight.

3. (optional) if u are implementing EAMSGD, you need to create a velocity array in the train_multi_machine for momentum and pass it to _update_params_on_kvstore.

4. modify train_multi_machine so u can call modified _update_params_on_kvstore according to the optimizer object type. Basically using "if else" stuff.

5. on each epoch testing, you need to synchronize each worker weight once using weight stored in kvstore(the averaging weight author used, I took the mean of works which is a little different). Make a copy of original one before this action,cause each worker weight is different, and restore it back after testing.

Questions:
Is there any documents explaining multi machine updating details? How my modification on _update_params_on_kvstore  and optimizer affects dist_sync case? 

Now I'm trying to scale this algorithm for multiple machines. My brain messed up with this case as I have no experience on this. I did 2 experiments using dmlc_mpi.py both with n=2, s=1 but 2 different kvstore setup (dist_sync, dist_async). the dist_async works while the dist_sync failed converging. I did't expected dist_async to work as my single machine code is for sync case, the single machine async case is not tested (because I don't know how to do so, LOL. I guess dmlc_local might be a direction).

**Comment by mli** [comment]
`_update_params_on_kvstore` means we will update the parameters by the optimizer set in the kvstore, namely by kvstore.set_optimizer(optimizer)

there are some documents explains the difference between dist_sync and dist_async

http://mxnet.dmlc.ml/how_to/multi_devices.html#distributed-training-with-multiple-machines
http://ps-lite.readthedocs.org/en/latest/overview.html

**Comment by Godricly** [comment]
Ye, I read the pslite and kv_store code and roughly understood how it works. I think my solution is functional but ugly. I used kv-store only to keep weight and do some synchronization. Thanks for the help.
